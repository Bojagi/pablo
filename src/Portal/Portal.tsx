import { forwardRef, ReactNode, useEffect, useState } from 'react';
import ReactDOM from 'react-dom';
import { setRef } from '../utils/setRef';

export interface PortalProps {
  children: ReactNode;
  name: string;
}

export const Portal = forwardRef<unknown, PortalProps>(({ children, name }, ref) => {
  const [mountPoint, setMountPoint] = useState<Element | null>(null);
  useEffect(() => setMountPoint(document.createElement('div')), []);
  useEffect(() => {
    if (mountPoint) {
      setRef(ref, mountPoint);
      mountPoint.setAttribute('data-testid', `pbl-${name}-mountpoint`);
      document.body.appendChild(mountPoint);
      return () => {
        document.body.removeChild(mountPoint);
      };
    }
    return () => {};
  }, [name, mountPoint, ref]);

  if (!mountPoint) {
    return null;
  }

  return ReactDOM.createPortal(children, mountPoint);
});
