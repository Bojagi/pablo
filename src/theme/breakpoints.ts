export type Breakpoint = 'sm' | 'md' | 'lg' | 'xl';

export const breakpointNames: Breakpoint[] = ['sm', 'md', 'lg', 'xl'];

export type BreakpointsArray = [number, number, number, number];

export type Breakpoints = BreakpointsArray &
  Record<Breakpoint, number> & {
    breakpointNames: Breakpoint[];
  };

export function createBreakpoints(
  array: BreakpointsArray,
  usedBreakpointNames: Breakpoint[] = breakpointNames
): Breakpoints {
  const newBreakpoints = ([...array] as unknown) as Breakpoints;
  usedBreakpointNames.forEach((bp, index) => {
    newBreakpoints[bp] = newBreakpoints[index];
  });
  newBreakpoints.breakpointNames = usedBreakpointNames;

  return newBreakpoints as Breakpoints;
}

export const defaultBreakpoints: BreakpointsArray = [600, 1000, 1200, 1920];

export const breakpoints = createBreakpoints(defaultBreakpoints, breakpointNames);
