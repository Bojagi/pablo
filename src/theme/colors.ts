export type ShadedColor = {
  main: string;
  dark: string;
  darkest: string;
  light: string;
  lightest: string;
  contrastText: string;
  contrastTextLight: string;
};

export type CommonColors = {
  white: string;
  black: string;
  whiteContrastText: string;
  blackContrastText: string;
};

export type TextColors = {
  main: string;
  info: string;
};

export type BorderColors = {
  main: string;
  lightest: string;
};

export interface GrayColors {
  '50': string;
  '100': string;
  '200': string;
  '300': string;
  '400': string;
  '500': string;
  '600': string;
  '700': string;
  '800': string;
  '900': string;
}

export type AllColors = ShadedColor & TextColors & BorderColors & CommonColors;

export interface Colors {
  common: CommonColors;
  gray: GrayColors;
  borders: BorderColors;
  brand: ShadedColor;
  positive: ShadedColor;
  negative: ShadedColor;
  neutral: ShadedColor;
  text: TextColors;
}

export const WHITE = '#ffffff';
export const BLACK = '#000000';

export const colors: Colors = {
  common: {
    white: WHITE,
    black: BLACK,
    whiteContrastText: BLACK,
    blackContrastText: WHITE,
  },
  gray: {
    '50': getGrayColor(0.95),
    '100': getGrayColor(0.9),
    '200': getGrayColor(0.8),
    '300': getGrayColor(0.7),
    '400': getGrayColor(0.6),
    '500': getGrayColor(0.5),
    '600': getGrayColor(0.4),
    '700': getGrayColor(0.3),
    '800': getGrayColor(0.2),
    '900': getGrayColor(0.1),
  },
  borders: {
    main: `${BLACK}${getHexByte(0.25)}`,
    lightest: `${BLACK}${getHexByte(0.1)}`,
  },
  text: {
    main: BLACK,
    info: `${BLACK}${getHexByte(0.5)}`,
  },
  brand: {
    lightest: '#EBE7FF',
    light: '#D6CEFF',
    main: '#6A50F2',
    dark: '#4129BD',
    darkest: '#281683',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  positive: {
    lightest: '#DFF6E6',
    light: '#BBECCB',
    main: '#38C765',
    dark: '#2FB85A',
    darkest: '#108B36',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  negative: {
    lightest: '#FFE7E5',
    light: '#FFCFCC',
    main: '#F21D0D',
    dark: '#CA271B',
    darkest: '#871A12',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  neutral: {
    lightest: '#FCF5DB',
    light: '#FFEB9E',
    main: '#FFDD5B',
    dark: '#EDC52A',
    darkest: '#D6AC0B',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
};

/**
 * Get hex color byte (e.g. to get transparency)
 * @param value Percentage value (between 0.0 and 1.0)
 */
function getHexByte(value: number) {
  return (Math.round((255 * value) / 16) * 16).toString(16);
}

function getGrayColor(value: number) {
  return `#${getHexByte(value)}${getHexByte(value)}${getHexByte(value)}`;
}
