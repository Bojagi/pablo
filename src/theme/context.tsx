import * as React from 'react';
import { PabloTheme, ComponentStyles } from './types';

export const pabloThemeContext = React.createContext<PabloTheme>({} as any);
export const pabloComponentStylesContext = React.createContext<ComponentStyles>({} as any);
export const rootContext = React.createContext<Document | ShadowRoot | null>(null);

export const useComponentStyleContext = () => React.useContext(pabloComponentStylesContext);

export const usePabloTheme = () => React.useContext(pabloThemeContext);
