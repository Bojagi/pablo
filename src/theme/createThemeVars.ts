import isObject from 'isobject';

type DeepStringReplace<T, M extends [any, any]> = {
  [P in keyof T]: T[P] extends object ? DeepStringReplace<T[P], M> : string;
};

export function createThemeVars<T extends Record<string, any>>(base: string, themeObject: T) {
  return internalThemeVars(`pbl-theme-${base}`, themeObject);
}

function internalThemeVars<T extends Record<string, any>>(
  base: string,
  themeObject: T
): DeepStringReplace<T, [number, string]> {
  return Object.entries(themeObject).reduce((acc, [key, value]) => {
    const keyVar = `${base}-${key}`;
    return { ...acc, [key]: isObject(value) ? internalThemeVars(keyVar, value) : `${keyVar}` };
  }, {} as any);
}

export function createThemeVarArray<T extends readonly string[]>(
  base: string,
  values: T
): Record<T[number], string> {
  return values.reduce(
    (acc, value) => ({ ...acc, [value]: `pbl-theme-${base}-${value}` }),
    {} as any
  );
}
