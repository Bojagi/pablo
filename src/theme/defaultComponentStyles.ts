import { buttonStyles, ButtonStyles } from '../Button/styles';
import { CardStyles, cardStyles } from '../Card/styles';
import { sidebarNavStyles, SidebarNavStyles } from '../SidebarNav/styles';
import { SwitchStyles, switchStyles } from '../Switch/styles';
import { TabsStyles, tabsStyles } from '../Tabs/styles';

export interface ComponentStyles {
  card: CardStyles;
  tabs: TabsStyles;
  sidebarNav: SidebarNavStyles;
  button: ButtonStyles;
  switch: SwitchStyles;
}

export const defaultComponentStyles: ComponentStyles = {
  card: cardStyles,
  tabs: tabsStyles,
  sidebarNav: sidebarNavStyles,
  button: buttonStyles,
  switch: switchStyles,
};
