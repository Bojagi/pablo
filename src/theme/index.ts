export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
