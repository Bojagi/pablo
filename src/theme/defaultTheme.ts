import { spacingTokens } from './spacing';
import { PabloTheme } from './types';
import { colors } from './colors';
import { typography } from './typography';
import { fluid } from './fluid';
import { breakpoints } from './breakpoints';

export const defaultTheme: PabloTheme = {
  spacing: spacingTokens,
  colors,
  typography,
  breakpoints,
  fluid,
};
