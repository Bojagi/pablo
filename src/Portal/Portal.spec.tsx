import {
  render,
  getByTestId as globalGetByTestId,
  getAllByTestId as globalGetAllByTestId,
  waitFor,
} from '@testing-library/react';
import React from 'react';
import { PabloThemeProvider, rootContext } from '../theme';
import { Portal } from './Portal';

test('Render Portal component', () => {
  const ref: any = {};
  const { container } = renderComponent({ ref });
  expect(container).toMatchSnapshot();
});

test('Forward ref object', () => {
  const ref: any = {};
  renderComponent({ ref });
  expect(ref.current).toBeDefined();
  expect(ref.current.getAttribute('data-testid')).toBe('pbl-portal-mountpoint');
});

test('Forward ref function', () => {
  const ref = jest.fn();
  renderComponent({ ref });
  expect(ref).toHaveBeenCalledTimes(1);
  expect(ref.mock.calls[0][0].getAttribute('data-testid')).toBe('pbl-portal-mountpoint');
});

test('Forward ref object with ShadowRoot', async () => {
  const ref: any = {};
  const shadowRoot = document.createElement('div').attachShadow({ mode: 'open' });
  renderComponent({ ref }, shadowRoot);
  expect(ref.current).toBeDefined();
  expect(ref.current.getAttribute('data-testid')).toBe('pbl-portal-mountpoint');
  expect(ref.current.parentElement).toBe(shadowRoot);
});

test('Forward ref function with ShadowRoot', () => {
  const ref = jest.fn();
  const shadowRoot = document.createElement('div').attachShadow({ mode: 'open' });
  renderComponent({ ref }, shadowRoot);
  expect(ref).toHaveBeenCalledTimes(1);
  expect(ref.mock.calls[0][0].getAttribute('data-testid')).toBe('pbl-portal-mountpoint');
});

function renderComponent(props: object, rootElement: ShadowRoot | Document = document) {
  console.log('document', rootElement);

  const { baseElement, unmount } = render(
    <PabloThemeProvider root={rootElement}>
      <Portal name="portal" {...props}>
        <div data-testid="outside-elem">Outside content</div>
      </Portal>
    </PabloThemeProvider>
  );

  return {
    getByTestId: globalGetByTestId.bind(null, baseElement),
    getAllByTestId: globalGetAllByTestId.bind(null, baseElement),
    unmount,
    container: baseElement,
  };
}
