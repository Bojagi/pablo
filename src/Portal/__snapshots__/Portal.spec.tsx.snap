// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render Portal component 1`] = `
<body>
  <div />
  <div
    data-testid="pbl-portal-mountpoint"
  >
    <div
      data-testid="outside-elem"
    >
      Outside content
    </div>
  </div>
</body>
`;
